cpp/src/arrow/types/primitive.h

```diff
@@ -261,10 +264,10 @@ typedef NumericBuilder<DoubleType> DoubleBuilder;
 
 class BooleanArray : public PrimitiveArray {
  public:
-  using PrimitiveArray::PrimitiveArray;
-
   BooleanArray(int32_t length, const std::shared_ptr<Buffer>& data,
       int32_t null_count = 0, const std::shared_ptr<Buffer>& null_bitmap = nullptr);
+  BooleanArray(const TypePtr& type, int32_t length, const std::shared_ptr<Buffer>& data,
```

Comment:
In theory it could have a different logical type (I can't think of a great example for boolean types).  I'm also including it here because we still haven't established if we want to be very pedantic about not doing operations that could fail with bad_alloc (especially in constructors).

Its a slightly bigger change but I can remove the requirement for passing this type and the the other ones on primitive.  I don't have a strong preference here.